WIP: Adapting the cfssl download to work also with the other architectures and improving the travis.yml

anagno opened the pull request:
This pull request implements:
* A new way of downloading the correct binary for each architecture for the cfssl library 
* Improving the the travis.yml. In the previous approach the UT were run in a after_success step (https://docs.travis-ci.com/user/job-lifecycle/#breaking-the-build). That step will NOT mark the build as failed if the UT fail. Moreover the deploy procedure is trigger now only in the case that the commits are coming from the osixia/docker-light-baseimage repo.

OLD Description:
While trying to make the docker-openldap a multi arch image, I realised that we were downloading a x86_64 binary that was not running on the aarch machines (e.g. raspberry). 
So this pull request implements a suggestion to solve this limitation.

With this pull request the source code of the cfssl library is being downloaded and compiled as a prebuild step. Then the produced binaries are included in the final image. (Currently I am using the official repo to download the source code, but that can be changed. I did not see any changes in the osixia repo, and I guessed that it is there only for creating and releasing the binaries).

Furthermore, I improved a little bit the travis.yml. In the previous approach the UT were run in a after_success step (https://docs.travis-ci.com/user/job-lifecycle/#breaking-the-build). That step will NOT mark the build as failed if the UT fail. Moreover the deploy procedure is trigger now only in the case that the commits are coming from the osixia/docker-light-baseimage repo.

I hope it is useful.
If something is not clear (or wrong), please let me know and I will try to explain it (fix it) :)

BertrandGouny commented:
this image aims to be as light as possible,
i'll prefer dowload cfssl when needed and add all cfssl bin available in https://github.com/osixia/cfssl releases. Not sure what is the best way to achieve that. Any help welcomed :)

anagno commented:
So the cfssl library is being download only and only if the user is using it. Did I get it right ?

If yes, then I can think a couple of choices:

* Download the source code and compile it on demand (this of course will need some time when being executed).
* use the uname -a (or the arch) command from bash to extract the correct ARCH and then download the correct binary from the repo (but of course we will have to precompile the binaries for all platforms -- Go supports that so it should not be problem).
* Use the debian package (https://packages.debian.org/sid/devel/golang-cfssl).
* Do not support cfssl on other architectures.

Which solution do you prefer/think it makes more sense?

KR,
Vasileios

P.S.: Disregard then also my PR in openldap image (https://github.com/osixia/docker-openldap/pull/333) :)

BertrandGouny commented:
thank you very mutch,
i think we should first try cfssl precompiled for all plateforms, it was pretty fast, i hope the binaries are ok :
https://github.com/osixia/cfssl/releases 

will try to adapt the download script accordingly

anagno commented:
If you prefer I can also do it (but probably over the weekend :) ) . Now that you upload the pre compiled binaries it is quite easy !

anagno committed:
Adding the .dockeringore file

anagno committed:
Updating the download script of the cfssl

This is necessary because the binary downloaded was a x86_64 and
it did not work on different arch images. So now we examine the
architecture of the system and download the correct binary.

anagno committed:
Improving the travis.yml

We now:
* Execute the tests after the build of the image is done,
and if the fail, we mark the build as failed
* We trigger the creation of the manifest after_deploy, otherwise
the build would me marked as failed even though the test passed

anagno commented:
This work in progress (sorry for the spam). Once it is ready I will leave a message :)

anagno committed:
Adding a work around for the arm build under travis

The reason for this workaround is documented in
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=923479

anagno commented:
Hello @BertrandGouny ,

I think this one is ready. I also tested it with the openldap image and with the exception of UT 7, which only succeeds in amd64 (e.g. https://travis-ci.org/anagno/docker-openldap/jobs/560681136#L1031) everything is green. I believe that reason behind this incompatibility is that the test files were generated in an amd64 machine. I will try and investigate further and find out the reason of this strange behaviour. What do you think?

KR,

Vasileios

anagno commented:
Introduced in 86392ca0c28e8776cd3c2dc28477b62b9fc8f484
